src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+                (actual, var remainingConnection) = ConsumeWindow(actual);
+
+                // Write what we can
+                if (actual < buffer.Length)
+                {
+                    buffer = buffer.Slice(0, actual);
+                }
+
+                var (hasMoreData, reschedule) = producer.Dequeue(buffer.Length, observed);
+
+                FlushResult flushResult = default;
+
+                if (readResult.IsCanceled)
+                {
+                    // Response body is aborted, complete reader for this output producer.
+                    if (producer.WriteHeaders)
+                    {
+                        // write headers
+                        WriteResponseHeaders(stream.StreamId, stream.StatusCode, Http2HeadersFrameFlags.NONE, (HttpResponseHeaders)stream.ResponseHeaders);
+                    }
+                }
+                else if (readResult.IsCompleted && stream.ResponseTrailers is { Count: > 0 } && !hasMoreData)
+                {
+                    // Output is ending and there are trailers to write
+                    // Write any remaining content then write trailers and there's no
+                    // flow control back pressure being applied (hasMoreData)
+
+                    stream.ResponseTrailers.SetReadOnly();
+                    stream.DecrementActiveClientStreamCount();
+
+                    // It is faster to write data and trailers together. Locking once reduces lock contention.
+                    flushResult = await WriteDataAndTrailersAsync(stream, buffer, producer.WriteHeaders, stream.ResponseTrailers);
+                }
+                else if (readResult.IsCompleted && producer.StreamEnded)
+                {
+                    if (buffer.Length != 0)
+                    {
+                        _log.LogCritical("{StreamId} observed an unexpected state where the streams output ended with data still remaining in the pipe.", stream.StreamId);
+                    }
+                    else
+                    {
+                        stream.DecrementActiveClientStreamCount();
+
+                        // Headers have already been written and there is no other content to write
+                        flushResult = await FlushAsync(stream, producer.WriteHeaders, outputAborter: null, cancellationToken: default);
+                    }
+                }
+                else
+                {
+                    var endStream = readResult.IsCompleted && !hasMoreData;
+
+                    if (endStream)
+                    {
+                        stream.DecrementActiveClientStreamCount();
+                    }
+
+                    flushResult = await WriteDataAsync(stream, buffer, buffer.Length, endStream, producer.WriteHeaders);
+                }
+
+                if (producer.IsTimingWrite)
+                {
+                    _timeoutControl.StopTimingWrite();
+                }
+
+                producer.WriteHeaders = false;
```

Comment:
Is there any chance that the first time a write gets enqueued `WriteHeaders` is false but then becomes true later? I don't think so, but this unsynchronized reading and setting of the field has me nervous.

Reply:
It isn't but I don't have a problem changing this state under a lock.

Reply:
Fixed